CRM/Contribute/Form/AdditionalPayment.php

```diff
@@ -527,20 +490,28 @@ public function processCreditCard($submittedValues) {
         $result = $payment->doPayment($paymentParams);
       }
       catch (\Civi\Payment\Exception\PaymentProcessorException $e) {
-        //set the contribution mode.
-        $urlParams = "action=add&cid={$this->_contactId}&id={$this->_id}&component={$this->_component}";
-        if ($this->_mode) {
-          $urlParams .= "&mode={$this->_mode}";
-        }
-        CRM_Core_Error::displaySessionError($result);
-        CRM_Utils_System::redirect(CRM_Utils_System::url('civicrm/payment/add', $urlParams));
+        Civi::log()->error('Payment processor exception: ' . $e->getMessage());
+        $errorMessage = $e->getMessage();
       }
     }
 
-    if ($result) {
-      $this->_params = array_merge($this->_params, $result);
+    // If payment is not completed then bounce approrptiate status message
+    $completedContributionStatusID = CRM_Core_PseudoConstant::getKey('CRM_Contribute_BAO_Contribution', 'contribution_status_id', 'Completed');
+    if (!empty($errorMessage) ||
+      (CRM_Utils_Array::value('payment_status_id', $result) != $completedContributionStatusID)
+    ) {
+      //set the contribution mode.
+      $urlParams = "action=add&cid={$this->_contactId}&id={$this->_contributionId}&component={$this->_component}&mode={$this->_mode}";
+      if (!$errorMessage) {
+        $errorMessage = ts('Payment from payment processor failed.');
+      }
+      CRM_Core_Session::singleton()->setStatus($errorMessage, ts('Payment Failed'), 'error');
```

Comment:
Any thoughts how this compares to using CRM_Core_Error::statusBounce from this line?